### DISCUSSION TITLE
Avoid manual type annotation for `useToggle`

### DISCUSSION THREAD

**Discussion by yudai-nkt** [mantinedev/mantine#1876] [Feature requests]
Mantine currently recommends explicitly setting a type parameter when using `useToggle`:

> By default, TypeScript will guess your type, but in most cases it's better to set type manually:
>
> ```ts
> const [value, toggle] = useToggle(['light', 'dark']); // value is string
> const [value, toggle] = useToggle<'dark' | 'light'>(['light', 'dark']); // value is 'dark' | 'light'
> ```

However, this is redundant and is not a very good idea IMHO. If I change the two values in the argument, the type parameter needs to be updated as well. I want to save this extra work and const assertion is the solution.

By improving the type definition and encouraging users to use const assertion instead of passing a type parameter, the variable `value` in the following code is inferred as `"light" | "dark"` ([playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDecCuAzkgMowoxJwC+ccAZlBCHAERSoasBQ3SAHpFgN8AOwzAIogsQAqEAOYKANkgA8sgHwAKCGBiTRhAFxwOKACZTlATzgBtWdlkBdAJS5udNFMLx7fhRI2MQwZEEucAC8MqTklLr6hoT2AAzuANy83r7wMIoqVDHaAG4oyvhIAPymyOgwAHQkSGHxSACCElIamh5Rmp50dMD0cNowNmBIEKNlFVQAhFExAORiFkj0wKJIFiseOF5DdKHhCXOVblnHtEjKxIPHcKdt2tpo+FAcojB9A4dPIYjMYfL5IH7RZZwPQGXxpdyPQF0DgwT7SGHJewARhc1yR1GySJRaOhSTh6TxNyuRzoBNpWSOxKg0gCbWw+SUqkiKEIcB8RhgWQJ3H5fgcF2CcA5hUiMSISHknKQ2nsrGUwAUAAsYKxsKwLCgoABrVjc3mi37cAD0VqeAD0qkA)).

```ts
import React, { useState }  from "react"

export function useToggle<T>(options: readonly [T, T]) {
  const [state, setState] = useState(options[0]);

  const toggle = (value?: React.SetStateAction<T>) => {
    if (typeof value !== 'undefined') {
      setState(value);
    } else {
      setState((current) => {
        if (current === options[0]) {
          return options[1];
        }

        return options[0];
      });
    }
  };

  return [state, toggle] as const;
}

const [value, toggle] = useToggle(["light", "dark"] as const);
```

What do you think about this? I'm ready to work on this feature once my proposal is accepted.

**Comment by rtivital** [comment]
Seems fine, feel free to create a PR

**Reply by yudai-nkt** [reply]
Done at #1892. I've just started using Mantine yesterday, and the experience has been great so far. So happy to contribute to this project!
